VladMeleshko/nodejs-course-2021Q2
# Task9/nestjs

1. Task: https://github.com/rolling-scopes-school/basic-nodejs-2021Q2/blob/master/descriptions/nestjs.md
2. Done: 11.07.2021 / deadline: 11.07.2021
3. Self-test ( 240 / 240 ):
[x] Guards должны использоваться для работы с авторизацией/аутентификацией. ( 30 / 30 )
[x] В приложении должны использоваться модули для разбиения структуры приложения на различные части (User, Board, Task). ( 30 / 30 )
[x] Exception filters должны использоваться для обработки исключений в приложении. ( 30 / 30 )
[x] @nestjs/typeorm должен использоваться для работы с базой данных. ( 30 / 30 )
[x] Для логирования может использоваться встроенный Logger или кастомная имплементация. ( 30 / 30 )
[x] Все внешние зависимости для модулей/классов должны предоставляться с помощью механизма dependency injection. ( 30 / 30 )
[x] В зависисимости от env переменной USE_FASTIFY Nest.js должен использовать или express или fastify. ( 30 / 30 )
[x] Необходимо сравнить производительность Nest.js с использованием express и fastify (можно использовать для этих целей artillery). ( 30 / 30 )

- Штрафов нет.
- Общая оценка - 240.

User with login='admin' and password='admin' is create in migrations, when you run `npm run start-migration` locally or automatically in docker (check instruction below).
If `USE_FASTIFY=true` in .env, Nestjs fastify will be used, else - Nestjs express.

How to check project local:

- verifie environment variables on .env;
- setup empty base created on .env;
- run `npm run start-migration`.

How to check project in docker:

- run `docker-compose up --build`.

If you first ran the application in docker, and then want to run it locally, start the server using command `npm run start-dev:bcrypt` or  `npm run start:bcrypt`.

## Test results:
### Express

|                      |                              |                                  |
| -------------------- | ---------------------------- | -------------------------------- |
| Scenarios            | [launched, completed]        | 300, 300                         |
| Requests             | [completed, response/sec]    | 1800, 27.11                      |
| Response time (msec) | [min, max, median, p95, p99] | 25, 5986, 1970.5, 4663.5, 5602.5 |
| Scenario counts      | [ratio]                      | 100%                             |
| Codes:               | [code:count]                 | 200: 1500, 201: 300              |

### Fastify

|                      |                              |                             |
| -------------------- | ---------------------------- | --------------------------- |
| Scenarios            | [launched, completed]        | 300, 300                    |
| Requests             | [completed, response/sec]    | 1800, 27.74                 |
| Response time (msec) | [min, max, median, p95, p99] | 9, 4595, 1644, 3958, 4273.5 |
| Scenario counts      | [ratio]                      | 100%                        |
| Codes:               | [code:count]                 | 200: 1500, 201: 300         |
